compiler/src/dmd/parse.d

```diff
@@ -8948,8 +8948,11 @@ class Parser(AST, Lexer = dmd.lexer.Lexer) : Lexer
                                 break;
                             check(TOK.comma);
                         }
-                        e = new AST.TupleLiteralExp(loc, elements);
                         check(loc, TOK.rightParenthesis);
+                        if (token.value == TOK.assign)
+                            e = new AST.TupleExp(loc, elements);
+                        else
+                            e = new AST.TupleLiteralExp(loc, elements);
```

Comment:
This looks like an unreliable way to do it, e.g.:
```d
( (a, b) ) = (1, 2);
```

Reply:
Also consider:
```d
(x ? (a, b) : (c, d)) = (1, 2);
```

Reply:
Also:
```d
((a, b), c) = ((1, 2), 3);
```

Reply:
```d
mixin(`(a, b)`) = t;
```

Reply:
How can we differentiate between lvalue tuple literals and rvalue ones in `ExpressionSemanticVisitor.visit(TupleLiteralExp)`?

Reply:
I am not sure what is the best way to do this in DMD short of actually keeping around the `TupleLiteralExp` until codegen.

Reply:
I had an idea that in itself won't make your examples work, but should make things easier. Instead of the `TupleLiteralExp` semantic call lowering it to a `CallExp`, it might work to make `TupleLiteralExp` derive from `CallExp`. Then we still get the lowering behaviour, but we can use a dynamic cast to determine if an expression is actually a tuple literal e.g. for `AssignExp` semantic. 

If that works, I think it will help in other ways for semantic analysis too, e.g. making a tuple literal not an lvalue (e.g. passing one as a `ref` argument should error).

Reply:
This seems like a bit of a hack. A tuple literal is not a `CallExp`, that's just the lowering strategy I employed to be able to run tests. It also cannot lower to Phobos imports in the long run. To be able to keep around the `TupleLiteralExp` in the AST, I think it is better to have the lowering as a field that is initialized during semantic.

Reply:
I'm looking into making *UnpackDeclaration* support unpacking into lvalue expressions. This would allow `int b; (auto a, b) = (1, 2);`.

Also, I think we can get rvalue errors for *TupleLiteralExp* by lowering it to `Tuple!Types(args)`. That should work because fields of a struct constructor call are still rvalues, unlike fields of a function call returning a struct (which are lvalues).

Reply:
> I think we can get rvalue errors

Didn't work. For `(5, 6)[0]++;` the lowering is:
```d
	((Tuple!(int, int) __tup24 = __tup24 = 0 , __tup24.__ctor(5, 6);) , __tup24.__expand_field_0)++;
```
Temporary `__tup24` being an lvalue means there's no rvalue error.

Reply:
Also `-vcg-ast` says `(a, b) = (3, 4);` generates:
```d
	Tuple(0, 0).__ctor(a, b).opAssign(Tuple(0, 0).__ctor(3, 4));
```
I don't know why that's not an rvalue error.

Reply:
You can call methods by reference on rvalues, so unfortunately that's what `opAssign` does.

Reply:
The alias this expansion resulting in an rvalue has also been a bit of a headache for `opArgs`.

Reply:
(E.g., some `isLvalue` helper in Phobos works by checking if a ref-arg function can be called, so it was true for all single-element tuples, which broke things.)

Reply:
Incidentally this is also the reason why `auto ref` is not supported.

Reply:
> You can call methods by reference on rvalues

Thanks, I've now documented that here:
https://github.com/dlang/dlang.org/pull/4235